grudge/dt_utils.py

```diff
@@ -0,0 +1,206 @@
+
+        # NOTE: order 0 elements have 1 node located at the centroid of
+        # the reference element and is equidistant from each vertex
+        if grp.order == 0:
+            assert nnodes == 1
+            min_delta_rs.append(
+                np.linalg.norm(
+                    nodes[0] - grp.mesh_el_group.vertex_unit_coordinates()[0]
+                )
+            )
+        else:
+            min_delta_rs.append(
+                min(
+                    np.linalg.norm(nodes[i] - nodes[j])
+                    for i in range(nnodes) for j in range(nnodes) if i != j
+                )
+            )
+
+    # Return minimum over all element groups in the discretization
+    return scaling * min(min_delta_rs)
+
+
+@memoize_on_first_arg
+def dt_geometric_factor(dcoll: DiscretizationCollection, dd=None) -> float:
+    r"""Computes a geometric scaling factor, defined as the minimum radius of
+    an inscribed circle/sphere in a mesh cell.
+
+    Specifically, the inradius for each element is computed using the following
+    formula for simplicial cells (triangles/tetrahedra):
+
+    .. math::
+
+        r_{in} = \frac{d V}{\sum_{i=1}^{N_{faces}} F_i},
+
+    where :math:`d` is the topological dimension, :math:`V` is the cell volume,
+    and :math:`F_i` are the areas of each face of the cell.
+
+    :arg dd: a :class:`~grudge.dof_desc.DOFDesc`, or a value convertible to one.
+        Defaults to the base volume discretization if not provided.
+    :returns: a :class:`float` denoting the geometric scaling factor.
+    """
+    from meshmode.discretization.poly_element import SimplexElementGroupBase
+
+    if dd is None:
+        dd = DD_VOLUME
+
+    actx = dcoll._setup_actx
+    volm_discr = dcoll.discr_from_dd(dd)
+
+    if any(not isinstance(grp, SimplexElementGroupBase)
+           for grp in volm_discr.groups):
+        raise NotImplementedError(
+            "Geometric factors are only implemented for simplex element groups"
+        )
+
+    cell_ones = volm_discr.zeros(actx) + 1.0
+    cell_vols = op.elementwise_sum(dcoll, op.mass(dcoll, dd, cell_ones))
+
+    # NOTE: The cell volumes are the *same* at each nodal location.
+    # Take the cell vols at each nodal location and average them to get
+    # a single value per cell.
+    @memoize_in(actx, (dt_geometric_factor, "cell_volume_knl"))
+    def cv_prg():
+        return make_loopy_program(
+            [
+                "{[iel]: 0 <= iel < nelements}",
+                "{[jdof]: 0 <= jdof < n_nodes}"
+            ],
+            """
+                result[iel] = sum(jdof, cell_vol[iel, jdof]) / n_nodes
+            """,
+            name="cell_volume"
+        )
+
+    cell_vols = DOFArray(
+        actx,
+        data=tuple(
+            actx.call_loopy(cv_prg(),
+                            cell_vol=cv_i)["result"]
+
+            for vgrp, cv_i in zip(volm_discr.groups,
+                                  actx.np.fabs(cell_vols))
+        )
+    )
```

Comment:
Could do this as an einsum with a post-multiply.

Reply:
Done: https://github.com/inducer/grudge/pull/108/commits/021fd898e939eebde9dc65e9700e9f9363a59f14